### PR TITLE
badger backend implementation

### PR DESCRIPTION
**Pull request by apesternikov**
this backend uses https://github.com/dgraph-io/badger

### CODE REVIEW

**Comment by mickael-carl** [comment]
Hi there! Thanks for contributing! I'm curious as to what the rationale behind using badger is. We already provide quite a few backends and I think it would be best to keep number of backends low, as this adds extra maintenance burden. Could you explain what would be the advantages compared to existing ones?

File: pkg/proto/configuration/blobstore/blobstore.proto [hunk 1]
```diff
@@ -190,3 +194,12 @@ message SizeDistinguishingBlobAccessConfiguration {
...
   int64 cutoff_size_bytes = 3;
 }
+
+message BadgerBlobAccessConfiguration {
+    // Directory where the files created by the badger storage
+    // backend are located.
+    string directory = 1;
+
+    // TTL for stored objects
+    google.protobuf.Duration ttl = 2;
```
Comment (mickael-carl) [inline pkg/proto/configuration/blobstore/blobstore.proto]: Could you document here the default value?

File: pkg/cas/byte_stream_server.go [hunk 2]
```diff
@@ -59,6 +60,7 @@ func NewByteStreamServer(blobAccess blobstore.BlobAccess, readChunkSize int) byt
 }
 
 func (s *byteStreamServer) Read(in *bytestream.ReadRequest, out bytestream.ByteStream_ReadServer) error {
+	log.Println("byteStreamServer.Read", in)
```
Comment (mickael-carl) [inline pkg/cas/byte_stream_server.go]: Shouldn't this make use of tracing?

File: pkg/blobstore/configuration/create_blob_access.go [hunk 3]
```diff
@@ -66,6 +67,17 @@ func createBlobAccess(configuration *pb.BlobAccessConfiguration, storageType str
 		return nil, errors.New("Storage configuration not specified")
 	}
 	switch backend := configuration.Backend.(type) {
+	case *pb.BlobAccessConfiguration_Badger:
+		backendType = "badger"
+		opts := badger.DefaultOptions
+		opts.Dir = backend.Badger.Directory
```
Comment (mickael-carl) [inline pkg/blobstore/configuration/create_blob_access.go]: Why not use `badger.DefaultOptions(backend.Badger.Directory)` ? According to [documentation](https://godoc.org/github.com/dgraph-io/badger#Options.WithDir) this sets both `Dir` and `ValueDir` to the provided path.

File: pkg/blobstore/configuration/create_blob_access.go [hunk 4]
```diff
@@ -66,6 +67,17 @@ func createBlobAccess(configuration *pb.BlobAccessConfiguration, storageType str
...
+		opts.ValueDir = backend.Badger.Directory
+		opts.SyncWrites = false
```
Comment (mickael-carl) [inline pkg/blobstore/configuration/create_blob_access.go]: Why use async writes? Isn't this a problem for cache consistency?

File: pkg/blobstore/badger_blob_access.go [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+	if ttl != nil {
+		ba.ttl = time.Duration(ttl.Seconds) * time.Second
+	}
+	go ba.gc()
+	return ba
+}
+
+// Badger storage garbage collection
+func (ba *badgerBlobAccess) gc() {
+	ticker := time.NewTicker(5 * time.Minute)
```
Comment (mickael-carl) [inline pkg/blobstore/badger_blob_access.go]: Out of curiosity, what is the reasoning behind choosing those values for garbage collection? Should they be configurable as well for end user?

File: pkg/blobstore/badger_blob_access.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+	if err := ctx.Err(); err != nil {
+		r.Close()
+		return err
+	}
+	value, err := ioutil.ReadAll(r)
+	r.Close()
+	if err != nil {
+		return util.StatusWrapWithCode(err, codes.Unavailable, "Failed to put blob")
+	}
+	span.Annotate(nil, "creating Transaction")
```
Comment (mickael-carl) [inline pkg/blobstore/badger_blob_access.go]: Nit: `Creating` rather than `creating`

File: pkg/blobstore/badger_blob_access.go [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+	txn := ba.db.NewTransaction(true)
+	defer txn.Discard()
+	span.Annotate(nil, "Transaction created")
+	err = txn.SetWithTTL([]byte(digest.GetKey(ba.blobKeyFormat)), value, ba.ttl)
+	if err != nil {
+		return util.StatusWrapWithCode(err, codes.Unavailable, "Failed to set blob")
+	}
+	span.Annotate(nil, "Set finished")
+	// Provide a callback so write is asynchronous.
+	err = txn.Commit(func(err error) {
```
Comment (mickael-carl) [inline pkg/blobstore/badger_blob_access.go]: As pointed out, async writes and deletion might break consistency, which is require, afaik.

File: pkg/blobstore/badger_blob_access.go [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+	if err := ctx.Err(); err != nil {
+		return nil, err
+	}
+	if len(digests) == 0 {
+		return nil, nil
+	}
+	txn := ba.db.NewTransaction(false)
+	defer txn.Discard()
+	var missing []*util.Digest
+	it := txn.NewIterator(badger.IteratorOptions{PrefetchValues: false})
```
Comment (mickael-carl) [inline pkg/blobstore/badger_blob_access.go]: From the [documentation](https://godoc.org/github.com/dgraph-io/badger#Txn.NewIterator): `Using prefetch is recommended if you're doing a long running iteration, for performance.`. What is the reason then to disable it?

**Comment by EdSchouten** [comment]
Closing this PR due to inactivity.

Note that I'm in principle not against adding support for more storage backends, do note that systems like Badger may not be a right fit. Log-structured merge-trees (LSMs) offer semantics that are far stronger than what is required by Buildbarn.

First of all, they are databases, not caches. This means that additional logic/processes needs to be added to constrain size. I think we can all agree that a cache should be simpler in terms of logic than a database. Using an LSM in combination with a garbage collector unfortunately is not.

Second, LSMs are designed to keep track of the key space in sorted order, so that it may be iterated. There is no need to apply any sorting in the case of the CAS and AC, as there simply is no iteration. Instead of an LSM, a hash table may be sufficient.

Since this PR was filed, Buildbarn has gained a new storage backend called LocalBlobAccess. I think time would better be spent either optimizing that one or adding any features that are missing.

**Pull request closed without merging**